Improve Operator Docs: Joining the Testnet (#126)
oasislabs/docs: src/operators/joining-the-testnet.md

:::

### Creating a Working Directory

During this entity initialization process, we will create keys and other
important artifacts that are necessary for the deployment of nodes on the
network. It is important that you save and protect the generated artifacts in
this directory if you intend to use them to register your entity and nodes.

::: tip NOTE
We will reference the working directory on `localhost` as `/localhostdir`
throughout the documentation.
:::

Inside `/localhostdir` you should create the following directories:

* `entity`: This will store your entity. The private contents in this directory
  are safest if used on a machine kept disconnected from the internet.

  The directory's permissions should be `rwx------`
* `node`: This will store a node we are calling "node". The name is not
  important. It simply represents one of your nodes. You can rename it to
  whatever you wish. The private contents of this directory will be used on the
  node itself.

  You should initialize this information on a system with access to the entity's
  private key.

  The directory permissions should be `rwx------`

To create the directory structure, use the following command:

```bash

> mkdir --mode 700 --parents {entity,node}

Reviewer: I'd rather not change this, as the old syntax works on both mac and linux.

Author: Oh, I didn't know about that.

Agreed, let's leave it as is then.

```

### Copying the Genesis File

> The latest genesis file can be found [Current Testnet Parameters][params].

Reviewer: ```suggestion
The latest genesis file can be found in [Current Testnet Parameters][params].
```

You should download the latest `genesis.json` file, copy it to the working
directory and set the following environment variable pointing to its path:

```bash
GENESIS_FILE_PATH=/localhostdir/genesis.json
```

This will be needed later when generating transactions.

### Initializing an Entity

As described in the [Architectural Overview][arch-entity], an entity is
critical to operating nodes on the network as it controls the stake attached to
a given individual or organization on the network.
In the future, we will support using Entity keys through HSMs to ensure that
entity keys cannot be easily compromised.

::: danger
We strongly suggest that you do not use any entity that is generated with the
current process on the Mainnet.

During the Public Testnet and staking competition we would also suggest that
you generate the entity on a system that has no network connection to provide
rudimentary protection for the entity key.
However, it is up to you to determine your own security practices.
:::

To initialize an entity simply run the following from `/localhostdir/entity`:

```bash
oasis-node registry entity init
```

> This will generate 3 files in `/localhostdir/entity`:

Reviewer: ```suggestion
This will generate three files in `/localhostdir/entity`:
```

### Copying the Genesis File

The latest genesis file can be found in [Current Testnet Parameters][params].
You should download the latest `genesis.json` file, copy it to the working
directory and set the following environment variable pointing to its path:

```bash

> GENESIS_FILE_PATH=/localhostdir/genesis.json

Reviewer 2: I would actually be weary of remove the `export` keyword only because it would make it easier for folks to copy paste. Tho I do understand why we wouldn't want it. I'm simply trying to manage the cost of removing that kind of thing and getting support requests for things where people don't actually export these variables but copy paste our multiline bash statements.

Author: Hmm... `export` should not be needed, even in multi-line Bash statements. For example:
```bash
[tadej@toronto ~]$ FOO=bar
[tadej@toronto ~]$ echo "This is a \
>   multi-line \
>   echo \
>   $FOO"
This is a   multi-line   echo   bar
[tadej@toronto ~]$
```

Reviewer 2: Ah yes... you are right since we're not using any scripts that rely on this. Ok! We're good :)

  --entity.node.descriptor node_genesis.json
```

This will update the entity descriptor in `entity.json` and subsequently the
`entity_genesis.json` file that contains the signed entity descriptor payload.

#### Initializing Additional Nodes

At the time of Public Testnet, the network will only have validators and no
other committees (no compute, no key management, no storage, etc.).

At this time this documentation does not include instructions for configuring
anything beyond a single validator.

If you'd like to create more validator nodes, you can simply repeat the process
above to initialize the artifacts for an additional node, just rename the
things pertaining to a particular node appropriately.

::: tip NOTE
Each node will require at least the network-defined minimum staking amount (at

> this time, this is 100 tokens).

Reviewer 2: @kostko I want to make sure this is actually true. I was under the impression that the minimum staking was a threshold minimum that wasn't additive per-node (because number of nodes is essentially irrelevant). I remember looking into this briefly with Peter.

Reviewer 3: >  I was under the impression that the minimum staking was a threshold minimum that wasn't additive per-node

The threshold is per node, based on role. E.g., each validator node requires at minimum X stake. Y validator nodes require at minimum Y*X stake. Currently there is no point of running more than one validator node per entity as only one can be scheduled at a time. But the thresholds behave as described.

:::

## Running an Oasis Node on the `server`

### Setting up the Oasis Node's Working Directory

Before we run the node on the `server` we need to ensure that we have a place
to store necessary files for the node.

::: tip NOTE

> We will reference the working directory on the `server` as `/serverdir`

Reviewer 2: :+1: :+1: much nicer :smile:

* `node/` - this is to store the node's data
* `node/entity/` - this is to store the public components of the node's entity

> You can make this directory structure by running the following command:

Reviewer 2: ```suggestion
You can make this directory structure by executing the following command:
``` perhaps this?

* `tls_identity.pem`
* `tls_identity_cert.pem`

After copying, make sure that all these files have `0600` permissions, i.e.
only their owner has `read` and `write` permissions.

To do so, run the following command:

```bash

> chmod --recursive 600 /serverdir/node/*.pem

Reviewer: This syntax doesn't work on mac, so would rather keep the existing version.

```

> ::: tip NOTE

Reviewer 2: ```suggestion
::: warning IMPORTANT NOTE
```

You may have noticed that some of these files were listed as **DO NOT SHARE**
in the [Initializing a Node] section.

In the future, these keys should be generated and referenced from an HSM.
However, until HSM support is implemented, these keys should be kept as secure
as possible on the `server`.
:::

[Initializing a Node]: #initializing-a-node

#### Copying the Public Entity Artifacts from `/localhostdir`

We will also need to have the public entity artifacts from the `/localhostdir`
present on the `server`.
Copy the `/localhostdir/entity/entity.json` file on `localhost` to
`/serverdir/node/entity/entity.json` on the `server`.

#### Copying the Genesis File to the server

The latest Genesis file can be found in the [Current Testnet Parameters][
params].
You should download the latest `genesis.json` file and copy it to
`/serverdir/etc/genesis.json` on the `server`.

#### Configuring the Oasis Node

::: warning NOTE
If you deployed a node on the 2019-11-13 Public Testnet, the configuration has
changed. Please, update your configuration or your node will fail to connect.
:::

::: tip NOTE

> 

Reviewer 2: Still not sure we should actually provide this instruction. Especially considering that turning it back on slows you down again. Can we leave it out this time? I think there might be more important things to fix before we can suggest this? Otherwise, to combat this issue we should have regular dump/restore cadences during the staking competition.

Author: I see your point. I've left this out here and [added a warning to the YAML config's comment](https://github.com/oasislabs/docs/pull/126/files#diff-f1db7a011b1d25cedbbcf9015e9cb3caR385-R394):
```yaml
tendermint:
  abci:
    prune:
      # NOTE: If you join the network at a later time, use the following
      # configuration to speed up the syncing of your node:
      #
      # strategy: none
      #
      # WARNING: Use this carefully. Turning pruning back on slows you down
      # again and not turning it on means you will eventually fill all the
      # storage on your node.
      # See https://github.com/oasislabs/oasis-core/issues/2432 for more
      # details.
      strategy: keep_n
      num_kept: 8640
```
If you think even this somewhat hidden comment is not appropriate, I can remove it.

If you are deploying a node after the Public Testnet has been running for more

> than a day, use the following Tendermint prunning configuration to speed up

Reviewer: ```suggestion
than a day, use the following Tendermint pruning configuration to speed up
```

your node syncing up with other nodes:

```yaml
tendermint:
  abci:
    prune:
      strategy: none
```

> In the future, we plan to speed up the prunning process so this will no longer

Reviewer: ```suggestion
In the future, we plan to speed up the pruning process so this will no longer
```

<!--
The <code v-pre> sections are due to an inability of vuepress to escape template
characters. We also had feedback that without the {{ }} surrounding the
variables the documentation was potentially ambiguous. Please keep the {{ }} in
the following section
-->

* <code v-pre>{{ external_address }}</code>: The external IP you used when
  registering this node.
  ::: tip NOTE
  If you are using a [Sentry Node], you should use the public IP of that
  machine.
* <code v-pre>{{ seed_node_address }}</code>:  The seed node address in the
  form `ID@IP:port`.

  You can find the current Oasis Seed Node address in the [Current Testnet
  Parameters][params].

> To you this configuration, save it in the `/serverdir/etc/config.yml` file and

Reviewer: ```suggestion
To use this configuration, save it in the `/serverdir/etc/config.yml` file and
```

      #
      # strategy: none
      strategy: keep_n
      num_kept: 86400
  core:
    listen_address: tcp://0.0.0.0:26656

    # The external IP that is used when registering this node to the network.
    # NOTE: If you are using the Sentry node setup, this option should be
    # omitted.
    external_address: tcp://{{ external_address }}:26656

  db:
    backend: boltdb
  debug:
    addr_book_lenient: false
  # List of seed nodes to connect to.
  # NOTE: You can add additional seed nodes to this list if you want.
  seed:
    - "{{ seed_node_address }}"
```

[Sentry Node]: ./sentry-node.md
[Oasis Core #2432]: https://github.com/oasislabs/oasis-core/issues/2432

#### Ensuring Proper Permissions

Only the owner of the process that runs the Oasis node should have access to
the files in the `/serverdir/node` directory.
The `oasis-node` binary ensures that the files used by the node are as least
privileged as possible so that you don't accidentally shoot yourself in the
foot while operating a node.

To ensure proper permissions are set, we suggest running the following to
remove all non-owner read/write/execute permissions:

```bash

> chmod --recursive go-r,go-w,go-x /serverdir

Reviewer: Similar to above, let's keep the existing syntax.

```

::: tip NOTE
Just so it's clear, the following permissions are expected by the `oasis-node`
binary:

* `700` for the `/serverdir/node` directory
* `700` for the `/serverdir/node/entity` directory
* `600` for all `*.pem` files
:::

### Starting the Oasis Node

::: danger WARNING
In a previous version of docs, we asked you to open port `42261` on a running
Docker container. In some configurations we noticed that this port was being
exposed to the outside world. This is no longer needed and should be removed
immediately. Keeping that port open was a temporary measure and is unsafe

> generally. Please, close that port to the public and do not bind to it in any

Reviewer: ```suggestion
generally. Please close that port to the public and do not bind to it in any
```

way.
:::

You can start the node by running the following command:

```bash
oasis-node --config /serverdir/etc/config.yml
```

> ::: tip

Reviewer 2: Can we add a word here? From a style perspective, we consistently use `NOTE`, `WARNING`, etc for other tips.

## Signing up for Testnet Tokens

::: tip NOTE
This step is not necessary if your entity is listed in the Genesis file or
already have tokens through some other means.

:::

In order to participate on the Testnet you'll need to have tokens. You'll use
these tokens to register your entity and stake on the network.

To get tokens, you'll need to sign up with [this form][tokens-form].
While filling out the form, it will ask for your entity's public key. This is
the same as your entity's account ID and can be found in the `id` key of the
`/localhostdir/entity/entity.json` JSON file.

> ::: tip EXAMPLE

Reviewer 2: Hrm... not 100% on this but I'm willing to try it. Mostly a personal style thing. Not yet sold on using the `tip` section for examples.

Author: I see your point. I'm not very tied to it, so just say so and I'll change it back :slightly_smiling_face:.

Reviewer 2: Hehe I'm also not certain. We can give it a shot. I appreciate that it calls attention to itself in a different way. So it's worth a shot.

## Signing up for Testnet Tokens

::: tip NOTE
This step is not necessary if your entity is listed in the Genesis file or
already have tokens through some other means.

> For most people, this will not be true.

Reviewer: ```suggestion
```

I think we just remove this line, as it is kind of confusing.

}
```

:::

After filling out the form, wait for an email notifying you that you've been
funded before moving forward. The following sections assume that you have
already been funded.

[tokens-form]: https://oasisfoundation.typeform.com/to/dlcekq

## Staking and Registering

::: tip NOTE

>  This step is not necessary if your entity is listed in the Genesis file.

Reviewer 2: We should include a note that you're in the `genesis` when we do a dump/restore.

Author: :+1:  I've added this explanation to the note.

}
```

:::

After filling out the form, wait for an email notifying you that you've been
funded before moving forward. The following sections assume that you have
already been funded.

[tokens-form]: https://oasisfoundation.typeform.com/to/dlcekq

## Staking and Registering

::: tip NOTE
 This step is not necessary if your entity is listed in the Genesis file.
:::

::: warning NOTE
If you've submitted staking or registry transactions before, your nonce is
likely different than the nonce used in the examples. If you're uncertain,
please check your account nonce by using [this guide][check-nonce].
:::

Once you have been funded, you can complete the process of connecting your

> node to the network by registering both, your entity and your node, as

Reviewer: ```suggestion
node to the network by registering both your entity and your node, as
```

> 

Reviewer: ℹ️ bookmark

### Generating a Staking (Escrow) Transaction on the `localhost`

Your entity's private key should be disconnected from the internet on the
`localhost`. Therefore, you need to generate the following transaction on the
`localhost`.

::: danger
The entity's private key is used to authorize transactions on your staking
account.

> Hence it should never be present on the `server`.

Reviewer: ```suggestion
Hence it should never be present on the network-connected `server`.
```

:::

For the Testnet, the current minimum stake required to register an entity and
register a node as a validator is 100 tokens.
So, we will generate an escrow transaction that escrows 100 tokens on your own
entity.

::: tip NOTE

> The Oasis node's staking application calls "staking" an escrow.

Reviewer: ```suggestion
The Oasis node's staking application calls the operation of staking tokens "escrow."
```
